fix: do not remove comments from XML during course import

Agrendalath opened the pull request:
This is a follow-up to edx#1087, which reverted this change.

According to the PR comments, parsing strings with XML comments inside them was causing errors. This does not seem to be the case anymore - these strings are just hidden when the block is rendered, but they are not breaking XBlocks.

## Jira

[OSPR-5996](https://openedx.atlassian.net/browse/OSPR-5996)
[BB-3827](https://tasks.opencraft.com/browse/BB-3827)

## Sandbox

https://pr28557.sandbox.opencraft.hosting/ (provisioning)


## Testing instructions
1. Create a new course with the following `Blank Advanced Problem`:
   ```xml
    <problem display_name="Checkboxes" markdown="null">
      <choiceresponse>
        <!-- This information is essential for the course authoring team. -->
        <p>
          Select a correct answer.
          <!-- Comment. -->
        </p>
        <checkboxgroup>
          <choice correct="true">a correct answer</choice>
          <choice correct="false">an incorrect answer</choice>
          <choice correct="false">an incorrect answer</choice>
        </checkboxgroup>
      </choiceresponse>
    </problem>
   ```

1. Export the course.
1. Import the exported course.
1. Check that the comments were retained in the XBlock.

## Deadline

None.

## Reviewers

- [x] @xitij2000 / @kaizoku 
- [ ] edX reviewer[s] TBD

**Settings**
```yaml
```

xitij2000 reviewed (approved):
:+1: 

- [x] I tested this: I tried exporting and importing a course without this PR and the XML comments were lost in the process. I tried again with this PR checked out and the comments are retained now.
- [x] I read through the code
- [na] I checked for accessibility issues
- [na] Includes documentation
- [na] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

Agrendalath commented:
@sarina, should we [treat](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529675973/Open+edX+Core+Committers#Reviews-and-Ownership) this as a bug fix, technical enhancement, or an end-user facing change? This is going to retain comments after importing the course, so technically it is going to face staff users, but not learners. Who could acknowledge/review this additionally?

sarina commented:
@Agrendalath I think I would treat as probably a technical enhancement. Since this is capa-related, the owning team is @edx/teaching-and-learning 

Would someone on T&L ack this change, as per https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529675973/Open+edX+Core+Committers#Reviews-and-Ownership which says that ack from owning team is sufficient for a PR with CC review. I tagged you because this changes the xml parser behavior.

sarina commented:
or possibly @nedbat who reviewed the original revert PR?

Agrendalath commented:
@sarina, won't we need another CC to review this change, as @xitij2000 is not the CC of `edx-platform` at the moment? Or does it still qualify as the non-author CC review?

sarina commented:
Ah, great point. Yeah it's be good if Ned or someone from T&L could review in that case.

nedbat commented:
The code I mentioned in my comment on the old PR is still in the repo (`process_xml(etree.tostring`).  It could be that XML comments in some particular places will cause problems, but the comments you used in the example here do not.

It is true that if you try to parse `"<!-- comment -->"` as XML, it will raise XMLSyntaxError.  There are nine occurrences of `process_xml(etree.tostring`.  Some catch exceptions and move on to the next child, which is fine.  Some do not.  LibraryContentBlock seems vulnerable.  Can you try adding an XML comment in a LibraryContentBlock?

Agrendalath commented:
Thank you for your guidance here, @nedbat. I've scheduled some time for the next week to verify this case and add a test for it.

Agrendalath commented:
@nedbat, you were right - when a comment was manually added to the LibraryContentBlock directly in the XML export, then this block was being ignored on import. I've changed this behavior (to ignoring XML comments) and added a test case for this.

I've also tried manually adding comments to verticals in the XML export and there were some exceptions raised by [modulestore/xml.py](https://github.com/edx/edx-platform/blob/8c0b4fed0e43650c3978f933604d5d28cf47cf23/common/lib/xmodule/xmodule/modulestore/xml.py#L161), but they are already handled in [seq_module](https://github.com/edx/edx-platform/blob/6fa89e610a679b76701d97d87c9bf65935622f38/common/lib/xmodule/xmodule/seq_module.py#L119) and [vertical_block](https://github.com/edx/edx-platform/blob/033497d1d7f58937d53b8b80ab859dd282cdb8ed/common/lib/xmodule/xmodule/vertical_block.py#L206).

I see that [this code](https://github.com/edx/edx-platform/blob/fb02353719419d7277d1453bf1a5fce5880d3625/common/lib/xmodule/xmodule/template_module.py#L166) from the TranslateCustomTagBlock might not handle this, but when I'm trying to add a `book`, `image`, or `slides` XBlock, I'm getting the following exception in the Studio. I've checked #26973, but I haven't found any additional steps required to set this up.
```python
Traceback (most recent call last):             
   File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/preview.py", line 327, in get_preview_fragment                                
     fragment = module.render(preview_view, context)
   File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/xblock/core.py", line 198, in render
     return self.runtime.render(self, view, context)
   File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1969, in render
     return self.__getattr__('render')(block, view_name, context)
   File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1418, in render
     return super().render(block, view_name, context=context)
   File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/xblock/runtime.py", line 823, in render                                              
     raise NoSuchViewError(block, view_name)
 xblock.exceptions.NoSuchViewError
```

nedbat commented on common/lib/xmodule/xmodule/library_content_module.py:
```diff
@@ -664,10 +665,20 @@ def get_content_titles(self):
+                    "Unable to load child when parsing Library Content Block. "
+                    "This can happen when a comment is manually added to the course export."
+                )
+                logger.error(msg)
```
This isn't an error, it would be better to do what XBlock does: explicitly check for a comment node and skip it.  But this will rarely happen, so let's move ahead.

nedbat commented on common/lib/xmodule/xmodule/tests/test_library_content.py:
```diff
@@ -74,18 +74,14 @@ class TestLibraryContentExportImport(LibraryContentTest):
     Export and import tests for LibraryContentBlock
     """
+    def setUp(self):
+        super().setUp()
```
You don't need to define setUp if it only calls super().setUp().  The parent's setUp will be inherited.

Agrendalath replied:
@nedbat, it looks a bit confusing in the diff, but both `super().setUp()` and `self.lc_block.refresh_children()` have the same indentation level.

nedbat commented on common/lib/xmodule/xmodule/tests/test_library_content.py:
```diff
@@ -74,18 +74,14 @@ class TestLibraryContentExportImport(LibraryContentTest):
+        self.lc_block = self.store.get_item(self.lc_block.location)
 
-        expected_olx = (
+        self.expected_olx = (
```
Why are these two variables changed to attributes?  They don't look like they are needed beyond the method.

nedbat commented:
Sorry for the delay.  This is good to merge.

nedbat reviewed (approved):
Let me know if you want me to merge, or if you want to make any more adjustments.

Agrendalath reviewed (commented):
@nedbat, I've squashed the commits. Please merge this when the CI passes. Thank you!

Agrendalath commented:
@nedbat, just a friendly reminder about this PR.

Agrendalath commented:
@nedbat, checking again, as it would be great to include this in Maple.

edx-webhook commented:
@jmbowman: thought you might like to know that Agrendalath merged this pull request.

<!-- comment:champion_merge_ping -->

edx-webhook commented:
@Agrendalath 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/28557&entry.1484655318=https://github.com/Agrendalath&entry.752974735=2021-08-26+15:36&entry.1917517419=2021-11-16+17:42&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

nedbat commented:
I've cherry-picked this onto Maple.

Agrendalath commented:
Thank you, @nedbat!
